### ISSUE TITLE
When using 13b, with the following configuration, a memory error occurs on v3-8. May I ask what is the reason?

### ISSUE THREAD

**Issue by joytianya** [young-geng/EasyLM#14]
2023-04-06 07:42:51.278258: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 7
2023-04-06 07:42:51.278359: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 1
2023-04-06 07:42:51.278403: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 3
2023-04-06 07:42:51.278439: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 2
2023-04-06 07:42:51.278486: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 0
2023-04-06 07:42:51.278526: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 4
2023-04-06 07:42:51.278553: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RE
SOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 5
2023-04-06 07:42:51.278591: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2153] Execution of replica 0 failed: RESOURCE_EXHAUSTED: Failed to allocate request for 625.00MiB (655360000B) on device ordinal 6

nohup python -m EasyLM.models.llama.llama_train \
    --mp_mesh_dim='-1,1' \
    --load_llama_config='13b' \
    --load_checkpoint="params::${EASYLM_CHECKPOINT_DIR}/checkpoint" \
    --tokenizer.vocab_file=${TOKENIZER_FILE} \
    --seed=42 \
    --initialize_jax_distributed=False \
    --total_steps=1000 \
    --log_freq=10 \
    --save_model_freq=100 \
    --save_milestone_freq=500 \
    --eval_steps=100 \
    --train_dataset.text_processor.fields='[input],output' \
    --train_dataset.text_processor.add_eos_token=True \
    --train_dataset.type='json' \
    --train_dataset.json_dataset.path=${TRAIN_DATA_FILE} \
    --train_dataset.json_dataset.seq_length=1024 \
    --train_dataset.json_dataset.batch_size=2 \
    --eval_dataset.text_processor.fields='[input],output' \
    --eval_dataset.text_processor.add_eos_token=True \
    --eval_dataset.type='json' \
    --eval_dataset.json_dataset.path=${EVAL_DATA_FILE} \
    --eval_dataset.json_dataset.seq_length=1024 \
    --eval_dataset.json_dataset.batch_size=2

**Comment by young-geng** [comment]
The memory on v3-8 is far from enough from training the 13b model. You'll need v3-64 at least.

**Comment by joytianya** [comment]
The total memory of v3-8 is 128GB, should that be enough?
and EasyLM is able to train large model that doesn't fit on a single accelerator by sharding the model weights.
If I misunderstood, could you please help explain it?

**Comment by joytianya** [comment]
What is the maximum batch size that v3-64 can support?

json_dataset.batch_size: does it refer to the batch size of a single core or the total batch size? 
Can easylm train script support accumulate_gradient_steps?

**Comment by young-geng** [comment]
Not sure about maximum batch size. I haven't tested that much with TPUs, as Koala was trained on Nvidia A100s.
The batch size is global and accumulate gradient can be configured in optimizer. I strongly recommend going through the documentation to understand how EasyLM works.

**Comment by joytianya** [comment]
thank you very much, I will read the document carefully again.

**Comment by joytianya** [comment]
The 13b llama weights would indeed only consume 52GB for float32, so 26gb for bf16/fp16?
Can using bf16 or fp16 training save half of the memory?

Does the training support specifying dtype?

**Comment by young-geng** [comment]
Currently EasyLM does not support pure bf16 training, as naive bf16 training could have some numerical problems.

**Comment by joytianya** [comment]
ok, so can easylm support float16 training?

**Comment by young-geng** [comment]
I currently don't have plans implementing that. Of course you are welcome to try implementing and testing that yourself.

**Comment by joytianya** [comment]
thank you ~

**Comment by s-smits** [comment]
> Currently EasyLM does not support pure bf16 training, as naive bf16 training could have some numerical problems.

Does it now? When SoTA pre-training methods are moving to int4 you'd except fp16 support at least.

**Comment by young-geng** [comment]
I think it should now. I've added the separation between params dtype and activation dtype in the flags. When training using bf16, you should set param_dtype to float32 and dtype to bf16.
